packages/contracts/src/RelayHub.sol

```diff
@@ -176,16 +180,21 @@ contract RelayHub is IRelayHub, Ownable {
     {
         RelayCallData memory vars;
         vars.initialGasLeft = aggregateGasleft();
+        vars.relayRequestId = GsnUtils.getRelayRequestID(relayRequest, signature);
         require(!isDeprecated(), "hub deprecated");
         vars.functionSelector = relayRequest.request.data.length>=4 ? MinLibBytes.readBytes4(relayRequest.request.data, 0) : bytes4(0);
-        require(msg.sender == tx.origin, "relay worker must be EOA");
-        vars.relayManager = workerToManager[msg.sender];
-        require(vars.relayManager != address(0), "Unknown relay worker");
-        require(relayRequest.relayData.relayWorker == msg.sender, "Not a right worker");
-        require(
-            isRelayManagerStaked(vars.relayManager),
-            "relay manager not staked"
-        );
+        if (msg.sender == batchGateway){
+            require(signature.length == 0, "batch gateway signature not zero");
+        } else {
+            require(msg.sender == tx.origin, "relay worker must be EOA");
```

Comment:
should add `require(signature.length!=0)`. 
Otherwise, anyone can send a request using Gatewayforwarder, on behalf of another sender

Reply:
Wider issue - Gateway must verify signature if it is set even coming from TrustedGateway, as otherwise RelayHub will have to verify signatures. Fixed (I think)

Reply:
where did you fix that?

Reply:
The new code looks like this:

```
        if (msg.sender != batchGateway){
            require(signature.length != 0, "missing signature or bad gateway");
            ...
            ...
```

so only BatchGateway can send a 0-length signature transactions.

Also in GatewayForwarder, the check is now

```
        if (msg.sender != trustedRelayHub || sig.length != 0) {
            super._verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
        }
```

so if the signature is set it will be verified.

My assumption is that it means there is no way to sneak a transaction that is not signed by the user through the GSN.

Am I correct?

Reply:
not sure what you mean by the assumption. anyone can create a transaction and call the relayHub (or gateway). our code should be resilient against such "rogue" transactions
we need to make sure that if it is sent directly to relayHub, it should fail (since the signature is not from user) and if it is sent via a gateway, it would fail, since the gateway will validate it (since it will fail the batch signature)

Reply:
My "assumption" is that `with the current code in this PR` there is no way to sneak a transaction that is not signed by the user through the GSN.